### PR TITLE
PlanTree Serialization

### PR DESCRIPTION
**Pull request by david-simoes-93**
Includes @bjoernschroeder 's changes to PlanInfoTree serialization, as well as a unit test for it with dynamic entrypoints (currently disabled)

### CODE REVIEW

File: alica_engine/src/engine/TeamObserver.cpp [hunk 1]
```diff
@@ -297,41 +302,41 @@ std::unique_ptr<SimplePlanTree> TeamObserver::sptFromMessage(AgentId agentId, co
...
+            // go up one level
             cur = curParent;
             if (cur == nullptr) {
-                ALICA_WARNING_MSG("TO: Malformed SptMessage from " << agentId);
-                return nullptr;
+                // went up one level from root node, finished parsing
+                break;
             }
             curParent = cur->getParent();
-            ++i;
```
Comment (veerajsk) [inline alica_engine/src/engine/TeamObserver.cpp:313]: don't we still need to increment i here?

File: alica_engine/src/engine/TeamObserver.cpp [hunk 2]
```diff
@@ -297,41 +302,41 @@ std::unique_ptr<SimplePlanTree> TeamObserver::sptFromMessage(AgentId agentId, co
...
+            continue;
+        }      
+        // Add node as child to old current node
+        const int64_t currentStateId = ids[i + 1];
+        if (i + 1 == static_cast<int>(ids.size())) {
```
Comment (veerajsk) [inline alica_engine/src/engine/TeamObserver.cpp:320]: this check should happen before the previous line

File: alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        // De-serialize (see TeamObserver::updateTeamPlanTrees and TeamObserver::handlePlanTreeInfo)
+        std::unique_ptr<SimplePlanTree> spi = ae->getTeamObserver().sptFromMessage(pti.senderID, pti.dynamicStateIDPairs, ae->getAlicaClock().now());
+        return spi;
+    }
+};
+
+/**
+ * Tests whether plan serialization works fine
+ */
+TEST_F(AlicaSerializationTest, serializeDeserialize)
```
Comment (veerajsk) [inline alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp]: need to add more comprehensive test cases. The tree should have atleast 2-3 levels with multiple children per node & try a few different tree structures

File: alica_tests/src/test/test_alica_dynamic_task.cpp [hunk 4]
```diff
@@ -149,5 +147,37 @@ TEST_F(AlicaDynamicTaskPlan, runDynamicTasks)
...
+
+        // Build current Plan Tree
+        IdGrp msg_to_send;
+        int deepestNode = 0;
+        int treeDepth = 0;
+        rpRoot->toMessage(msg_to_send, rpRoot, deepestNode, treeDepth);
+        plan_trees[agent_index] = msg_to_send;
+    }
+    // TODO
+    // ASSERT_NE(plan_trees[0], plan_trees[1]);
```
Comment (veerajsk) [inline alica_tests/src/test/test_alica_dynamic_task.cpp:180]: need to assert if the plan tree is published as expected i.e. the structure + <stateid, dynamic id>. Sucess marks can be checked later in a different test case.

Comment (david-simoes-93) [reply alica_tests/src/test/test_alica_dynamic_task.cpp:180]: Test isn't passing still, so I left it commented until feature is implemented

File: alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp (hunk 3 above)
Comment (david-simoes-93) [reply alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp]: Test isn't passing still, so I left it commented until feature is implemented

File: alica_engine/src/engine/TeamObserver.cpp (hunk 1 above)
Comment (david-simoes-93) [reply alica_engine/src/engine/TeamObserver.cpp:313]: @bjoernschroeder Can you address this one?

File: alica_engine/src/engine/TeamObserver.cpp (hunk 2 above)
Comment (david-simoes-93) [reply alica_engine/src/engine/TeamObserver.cpp:320]: @bjoernschroeder Can you address this one?

File: alica_engine/src/engine/TeamObserver.cpp (hunk 1 above)
Comment (veerajsk) [reply alica_engine/src/engine/TeamObserver.cpp:313]: @bjoernschroeder

File: alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp (hunk 3 above)
Comment (veerajsk) [reply alica_tests/src/test/test_alica_simple_plan_tree_serialization.cpp]: yeah, I don't think the deserialization will work as it is right now

File: alica_engine/src/engine/TeamObserver.cpp (hunk 1 above)
Comment (bjoernschroeder) [reply alica_engine/src/engine/TeamObserver.cpp:313]: No I dont think we need to increment here. After -1, the next id will be the currentEPId followed by the currentStateId. If you have a list like that:

[..., ep_0, s_0, -1, ep_1, s_1]
with ep_i is the i-th entrypoint id and s_i is the i-th state id.

If the value at the current index is -1, we can just continue with the next entry in the list.

EDIT: Just saw that the for loop does not increment with each iteration, so we have to always manually increment it. I will add the increment.

**Review by veerajsk — APPROVED** [review]

**Pull request merged**